File: Minimum Absolute Difference in BST - Leetcode 530/Minimum Absolute Difference in BST - Leetcode 530.py
class Solution:
    def getMinimumDifference(self, root: Optional[TreeNode]) -> int:
        min_distance = [float('inf')]
        prev = [None]

        def dfs(node):
            if node is None:
                return
                
            dfs(node.left)

            if prev[0] is not None:
                min_distance[0] = min(min_distance[0], node.val - prev[0])

            prev[0] = node.val
            dfs(node.right)

        dfs(root)
        return min_distance[0]
        # Time: O(n)
        # Space: O(n)

File: Minimum Absolute Difference in BST - Leetcode 530/Minimum Absolute Difference in BST - Leetcode 530.js
var getMinimumDifference = function(root) {
    let minDifference = Infinity;
    let prev = null;

    function dfs(node) {
        if (node === null) return;

        dfs(node.left);

        if (prev !== null) {
            minDifference = Math.min(minDifference, Math.abs(node.val - prev));
        }
        prev = node.val;

        dfs(node.right);
    }

    dfs(root);
    return minDifference;
};
